src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,544 @@
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            IEqualityComparer<TKey> comparer = _comparer;
+            hashCode = (comparer?.GetHashCode(key) ?? key.GetHashCode()) & 0x7FFFFFFF;
+            int[] buckets = _buckets;
+            int index = -1;
+            if (buckets.Length > 0)
+            {
+                int bucket = hashCode % buckets.Length;
+                index = buckets[bucket] - 1;
+                comparer = comparer ?? EqualityComparer<TKey>.Default;
+                Entry[] entries = _entries;
+                while (index >= 0)
+                {
+                    Entry entry = entries[index];
+                    if (entry.HashCode == hashCode && comparer.Equals(entry.Key, key))
+                    {
+                        break;
+                    }
+                    index = entry.Next;
+                }
+            }
+            return index;
+        }
+
+        private bool TryInsert(int? index, TKey key, TValue value, InsertionBehavior behavior)
+        {
+            int i = IndexOf(key, out int hashCode);
+            Entry[] entries = _entries;
+            if (i >= 0)
+            {
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        entries[i].Value = value;
+                        return true;
+                    case InsertionBehavior.ThrowOnExisting:
+                        throw new ArgumentException(string.Format(Strings.Argument_AddingDuplicateWithKey, key.ToString()));
+                    default:
+                        return false;
+                }
+            }
+
+            ++_version;
+            // Check if resize is needed
+            int count = Count;
+            if (entries.Length == count)
+            {
+                Resize(HashHelpers.ExpandPrime(entries.Length));
+                entries = _entries;
+            }
+
+            // Increment indices >= index;
+            int actualIndex = index ?? count;
+            for (i = actualIndex; i < count; ++i)
+            {
+                UpdateBucketIndices(i, 1);
+            }
+            Array.Copy(entries, actualIndex, entries, actualIndex + 1, count - actualIndex);
+
+            ref int bucket = ref _buckets[hashCode % _buckets.Length];
+            Entry newEntry = new Entry { HashCode = hashCode, Key = key, Value = value, Next = bucket - 1 };
+            entries[actualIndex] = newEntry;
+            bucket = actualIndex + 1;
+            ++Count;
+            return true;
+        }
+
+        private void RemoveEntryFromBucket(int i)
+        {
+            Entry[] entries = _entries;
+            Entry entry = entries[i];
+            ref int b = ref _buckets[entry.HashCode % _buckets.Length];
+            if (b == i + 1)
+            {
+                b = entry.Next + 1;
```

Comment:
`// Bucket was pointing to removed entry. Update it to point to the next in the chain` ?